cpp/include/xtt/server_handshake_context.hpp

```diff
@@ -0,0 +1,109 @@
+/******************************************************************************
+ *
+ * Copyright 2018 Xaptum, Inc.
+ * 
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ * 
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+
+#ifndef XTT_CPP_SERVERHANDSHAKECONTEXT_HPP
+#define XTT_CPP_SERVERHANDSHAKECONTEXT_HPP
+#pragma once
+
+#include <xtt/context.h>
+#include <xtt/messages.h>
+
+#include <xtt/pseudonym.hpp>
+#include <xtt/longterm_key.hpp>
+#include <xtt/identity.hpp>
+#include <xtt/group_identity.hpp>
+#include <xtt/group_public_key_context.hpp>
+#include <xtt/types.hpp>
+#include <xtt/server_certificate_context.hpp>
+#include <xtt/server_cookie_context.hpp>
+
+#include <memory>
+#include <experimental/optional>
+
+namespace xtt { class server_handshake_context; }
+void swap(xtt::server_handshake_context&, xtt::server_handshake_context&);
+
+namespace xtt {
+
+    class server_handshake_context {
+    public:
+        struct io_buffer {
+            unsigned char* io_ptr;
```

Comment:
Maybe:

- `io_ptr` to `data` or `ptr`
- `io_bytes_requested` to `size` or `len`

Just to keep the call sites more readable:  `async_read(buf.ptr, buf.size)` instead of `async_read(buf.io_ptr, buf.io_bytes_requested`).